### ISSUE TITLE
adviving/aliasing of functions

### ISSUE THREAD

**Issue by ksjogo** [ksjogo/emacs-libgit2#8]
Shall we defadvice/defalias/some other way patch the external functions we want to make faster?

**Comment by vermiculus** [comment]
I doubt it's advantageous to allow customization *within* a C function, but I see no reason (or ability) to stop folks from advising/aliasing them.

This does raise an interesting question though: do we wish to do the higher-level programming in C, or should we stick to just implementing the API in C and exposing them as Lisp objects so the higher-level stuff can be done in Lisp? I think performance would be comparable, but that would require actual testing, of course.

**Comment by ksjogo** [comment]
Ok, misworded the question a bit, how should we approach injecting ourselves into magit/gutter etc.? This leads to your question, what is the best return format for them at which point. Gutter seems to be writing git command output into a buffer and then parsing that, which seems to be wasteful and unnecessary for us, as we had to emulate the standard git style and do an unnecessary encode/decode step.
I suppose, we could write some elisp constructor functions returning value objects, which we then call from our c code, as simple return values would require a lot of calling into the lib.

**Comment by vermiculus** [comment]
> How about using a feature branch in the magit repository?

That was my assumption as well, but see melpa/mepla#4321.

**Comment by ksjogo** [comment]
Multiple repos would have the disadvantage of requiring the development sync for everything. If we just keep stuff in here for the moment, we can test against that and have no cross hassle. 
Though for later it is probably a lot better.

**Comment by tarsius** [comment]
>> How about using a feature branch in the magit repository?

> That was my assumption as well, but see melpa/melpa#4321.

I fail to see how these two things are related, but see my comment on that issue.

**Comment by tarsius** [comment]
> Multiple repos would have the disadvantage of requiring the development sync for everything. If we just keep stuff in here for the moment, we can test against that and have no cross hassle.

Agreed. Though at this early stage you don't really have to teach Magit how to use libgit in order to improve libgit. I think you should instead implement support for some features and "use" them in tests rather than patching/advising Magit.

Once you have a set of features you can start using them in other packages. And once you get to that point it will be much easier for the maintainers of those packages to figure out what has changed if they can look at a diff instead of at advise definitions.

But if you want to start using libgit in Magit as soon as possible, then go ahead and use advices. You will end up having to redo that work though.

**Comment by vermiculus** [comment]
> I fail to see how these two things are related, but see my comment on that issue.

@tarsius sorry, I meant to add that if magit starts depending on libgit2, installing magit over melpa would need to pull in libgit2 as a dependency (presumably also over melpa) and we'll have to figure out how that's going to work since it C/Rust are compiled languages. That's what the linked issue is about.

All of this is pretty far off though – it'll be at *least* a few months until this is ready for primetime, I think.
